dali/test/dali_operator_test.h

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_TEST_DALI_OPERATOR_TEST_H_
+#define DALI_TEST_DALI_OPERATOR_TEST_H_
+
+#include <gtest/gtest.h>
+#include <vector>
+#include "tensor_adapter.h"
+#include "operators_graph.h"
+
+namespace dali {
+
+namespace testing {
+
+template<typename InputType, typename OutputType>
+class DaliOperatorTest : public ::testing::Test {
+  static_assert(std::is_fundamental<InputType>::value, "DaliOperatorTest expects fundamental type as InputType");
+  static_assert(std::is_fundamental<OutputType>::value, "DaliOperatorTest expects fundamental type as OutputType");
+
+public:
+  using Arguments =  std::map<std::string, int>;  // TODO some generalization (boost::any? tagged union?)
+
+  void RunTest(Arguments operator_arguments, std::vector<TensorAdapter<OutputType>> anticipated_outputs) noexcept {}
+
+private:
+  virtual std::vector<TensorAdapter<InputType>> SetInputs() const noexcept = 0;
+
+  virtual OperatorsGraph SetOperators() const noexcept = 0;
+
+  virtual bool Verify(TensorAdapter<OutputType> output, TensorAdapter<OutputType> anticipated_output) const noexcept = 0;
```

Comment:
1. This hides the output tensor object from test developer.
2. It doesn't handle multiple outputs.
3. The function returns `bool`. While having this output value is nice, it's very easy to (mis)understand as a a binary predicate - it just looks at the input and output and says "good" or "bad" without any further insights. We should make it clear, somehow, that this function should print some kind of diagnostic _or_ return a descriptive error message.
I thought of something like this:
```
using Verify = ErrorDesc(
    const std::vector<TensorList *> &output, // pointers to actual tensors returned by the operator in the pipeline
    const std::vector<TensorList> &referenceOutput
);

std::function<Verify> verify = dali::testing::TensorsEqual;
```

Reply:
After consideration, I agree, that single return value removes possibilities of quite powerful GTest features. Thus, I think that developer here should provide himself calls for GTest-checkers (`ASSERT_TRUE`, `EXPECT_FLOAT_EQ`, etc.)

However, I still don't see reason, why providing here `dali::TensorList<>` is better, than having an adapter for TensorList. Such adapter has all the attributes TensorList has, provides easy way for converting and in case TensorList changes, we change only one function, not every test.